docs(docs-infra): Add cross-links to API pages

JeanMeche opened the pull request:
This commit adds links to our own symbols in the API pages, including the Code TOC, the description and the Usage notes. 

This commit also add the implement/extends symbols for classes & interfaces. 

The generation process now logs warnings for links that couldn't not be generated because the symbols can be resolved in the given context. 


fixes https://github.com/angular/angular/issues/57128
fixes https://github.com/angular/angular/issues/57183
fixes #55571


Worthy examples: 
* https://ng-dev-previews-fw--pr-angular-angular-57346-adev-prev-ckp3dlh0.web.app/api/core/afterRender
* https://ng-dev-previews-fw--pr-angular-angular-57346-adev-prev-ckp3dlh0.web.app/api/common/LocationStrategy?tab=description
* https://ng-dev-previews-fw--pr-angular-angular-57346-adev-prev-ckp3dlh0.web.app/api/common/NgForOfContext
* https://ng-dev-previews-fw--pr-angular-angular-57346-adev-prev-ckp3dlh0.web.app/api/common/HashLocationStrategy

JeanMeche commented on adev/shared-docs/pipeline/api-gen/rendering/processing.ts:
```diff
@@ -39,38 +39,42 @@ import {getTypeAliasRenderable} from './transforms/type-alias-transforms';
   moduleName: string,
+  symbols: SymbolMap,
 ): DocEntryRenderable | CliCommandRenderable {
   if (isClassEntry(entry)) {
```
Thought: Should `symbols` be a scoped global so we don't need to passed all the way down ?

devversion commented on adev/shared-docs/pipeline/api-gen/extraction/index.ts:
```diff
@@ -54,12 +54,20 @@ function main() {
   const combinedEntries = extractedEntries.concat(extraEntries);
 
+  // Add the module symbols also to the map
+  const strippedModuleName = moduleName.replace('@angular/', '');
```
I think this shows (for the other PR with the core/globals) that we need to maybe have `module_name`, and a "title" field? I'm worried we are using `module_name` for more, but then `Dev-mode globals` breaks

JeanMeche replied:
Yes, we'll be able to reuse the logic that I added in #57365

devversion commented on adev/shared-docs/pipeline/api-gen/rendering/entities.ts:
```diff
@@ -205,6 +205,8 @@ export interface InitializerApiFunctionEntry extends DocEntry {
   };
 }
 
+export type SymbolMap = Map<string, string>;
```
Nit: Add description

devversion commented on packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:
```diff
@@ -62,6 +66,11 @@ export class DocsExtractor {
         // We want the real source file of the declaration.
         const realSourceFile = node.getSourceFile();
 
+        const importedSymbols = this.getImportedSymbols(realSourceFile);
```
This computation happens for every export. I also wonder, why is this necessary? can you explain? i.e. why do we need to extract the "imported" symbols and returning them as `symbols` here is a bit ambiguous I feel like.

JeanMeche replied:
I've added a comment, let me know if it makes it more clear for you.

devversion commented:
I like the general approach, but I also don't know how it was done before in the early versions of angular.dev, or why we stopped/reverted the logic, so I will defer to @jelbourn on this for now.

josephperrott commented:
Deferring to other reviewers for this specific change.

JeanMeche commented:
Currently trying to reduce the scope of this PR, as it might feel a bit large. 
I think the important part are the `compiler-cli` changes.

jelbourn commented on packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:
```diff
@@ -30,6 +30,8 @@ import {extractTypeAlias} from './type_alias_extractor';
 
 type DeclarationWithExportName = readonly [string, ts.Declaration];
 
+const privateModules = new Set(['core/primitives/signals']);
```
Needs a comment explaining what this is and why it's necessary

(in general, I'm looking for comments to tell me _why_ the code does something over explaining _what_ the code does)

jelbourn commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -902,7 +903,7 @@ export class NgCompiler {
    *     docs should be extracted.
    */
-  getApiDocumentation(entryPoint: string): DocEntry[] {
+  getApiDocumentation(entryPoint: string): {entries: DocEntry[]; symbols: SymbolMap} {
```
I'd want to create a full type for this return value with appropriate JsDoc. Just `symbols` by itself, too, is a bit vague. Maybe something like `importedSymbols` or `allImportedSymbols`

jelbourn commented on packages/compiler-cli/src/ngtsc/docs/src/class_extractor.ts:
```diff
@@ -343,6 +378,15 @@ class ClassExtractor {
+  /**
+   * Check wether a member has a private computed property name like [ɵWRITABLE_SIGNAL]
+   */
+  private hasPrivateComputedProperty(property: PropertyLike) {
```
Needs a comment that explains _why_ this is necessary

jelbourn commented on packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:
```diff
@@ -136,6 +156,40 @@ export class DocsExtractor {
+  /**
+   * @returns the symbols from other @angular packages
+   */
+  private getImportedSymbols(sourceFile: ts.SourceFile): SymbolMap {
```
I would break this out into its own extractor file

jelbourn reviewed (commented):
I got part of the way through before I need to run for the day, but here's my initial comments

Not sure if you were planning to do this already once you got an okay on the approach, but I'll mention anyway: we should split this change up into more granular commits. In particular:
* Separating the symbol collection into its own commit (and this also needs tests)
* Changing the representation of class inheritance in its own commit
* Minimum necessary changes to the renderer to accomodate any `DocEntry` structure changes
* Further changes to surface the links in the rendered output

JeanMeche committed:
refactor(compiler-cli): Add a map of every symbols used inside a package

This commit changes the structure of the API extraction files to include all symbols used inside a package.

The structure is a `Map`, Symbol => package
eg: 'ApplicationRef' => '@angular/core'

JeanMeche committed:
docs(docs-infra): Add support for cross-links on API pages

JeanMeche commented:
I splitted the changes accross multiple commits, I should ease the review of all the changes (or we can spin them off in other PRs)

JeanMeche commented on adev/shared-docs/pipeline/api-gen/extraction/index.ts:
```diff
@@ -54,12 +54,20 @@ function main() {
+      ...apiDoc.symbols.entries(),
+
+      // Exported symbols from the current package
+      ...apiDoc.entries.map((entry) => [entry.name, moduleName]),
```
@jelbourn Do you think we should do this concatenation during extraction ?

jelbourn reviewed (approved):
Overall looks good

JeanMeche commented on adev/shared-docs/pipeline/api-gen/rendering/transforms/code-transforms.ts:
```diff
@@ -428,3 +435,28 @@ function appendPrefixAndSuffix(entry: DocEntry, codeTocData: CodeTableOfContents
+ */
+export function addApiLinksToHtml(htmlString: string): string {
+  const result = htmlString.replace(
+    /(?<!<a[^>]*>)(<(?:(?:span)|(?:code))[^>]*>\s*)([^<]*?)(\s*<\/(?:span|code)>)/g,
```
It is valid, make sure that the visualizer supports ES regex, like https://regex101.com/

devversion commented on packages/compiler-cli/src/ngtsc/docs/src/import_extractor.ts:
```diff
@@ -0,0 +1,45 @@
+    if (ts.isImportDeclaration(node)) {
+      let moduleSpecifier = node.moduleSpecifier.getText(sourceFile).replace(/['"]/g, '');
+
+      if (moduleSpecifier.startsWith('@angular')) {
```
```suggestion
      if (moduleSpecifier.startsWith('@angular/')) {
```

devversion replied:
would it be better/more generic if we filter everything that is not a relative path?

JeanMeche replied:
That would also expose symbols from RxJS, which is not wanted.

devversion replied:
It might be the responsibility of the "consumer" of the symbols list to filter out items that cannot be rendered. Or even cooler, we could link to the RxJS docs that way 😄

devversion commented on packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:
```diff
@@ -62,6 +70,20 @@ export class DocsExtractor {
         const realSourceFile = node.getSourceFile();
 
+        /**
+         * `sourceFile` is generaly a `export * from './public_api';` with no imports.
```
```
         * The `sourceFile` from `extractAll` is the main entry-point file of a package.
         * Usually following a format like `export * from './public_api';`, simply re-exporting.
```

devversion commented on packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:
```diff
@@ -62,6 +70,20 @@ export class DocsExtractor {
 
+        /**
+         * `sourceFile` is generaly a `export * from './public_api';` with no imports.
+         * It is necessary to pick-up every import from the real source files.
```
```
         * It is necessary to pick-up every import from the actual source files
         * where declarations are living, so that we can determine what symbols
         * are actually referenced in the context of that particular declaration
```

devversion commented on packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:
```diff
@@ -62,6 +70,20 @@ export class DocsExtractor {
+         * `sourceFile` is generaly a `export * from './public_api';` with no imports.
+         * It is necessary to pick-up every import from the real source files.
+         * This allows to include symbols from other packages (like @angular/core)
+         * By doing this, the generation remains independant from other packages
```
```suggestion
         * By doing this, the generation remains independent from other packages
```

devversion commented on adev/shared-docs/pipeline/api-gen/rendering/symbol-context.ts:
```diff
@@ -0,0 +1,18 @@
+ * eg: 'ApplicationRef' => 'core', 'FormControl' => 'forms'
+ */
+export function getModuleName(symbol: string): string | undefined {
+  return symbols.get(symbol)?.replace('@angular/', '');
```
Should `symbol` in the extractor simply have the "package name" already? It feels like we are spreading all these cutoffs throughout various places.

JeanMeche replied:
With the latest refactors, it's the only place where we do it (we also don't it anymore in #57365)

devversion commented on adev/shared-docs/pipeline/api-gen/extraction/index.ts:
```diff
@@ -54,12 +54,20 @@ function main() {
     moduleName: moduleName,
     entries: combinedEntries,
+    symbols: [
+      // Imported symbols from other packages
```
```suggestion
      // Symbols referenced, originating from other packages
```

devversion commented on packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:
```diff
@@ -62,6 +70,20 @@ export class DocsExtractor {
+          if (privateModules.has(moduleName) || symbolName.startsWith('ɵ')) {
+            return;
+          }
+          symbols.set(symbolName, moduleName);
```
Is this logic 100% correct? Imagine two declarations from different files import a different `ApplicationRef`. The resolved imports would be different, from different packages potentially? 

If so, please also elaborate this in a comment for future-proofness.

JeanMeche replied:
This would indeed be an issue. 

I think the only existing case today is for `inject` which is exposed by both `core` and `core/testing` (we had that same issue on angular.io). 

We could either have an array in the map or drop the entry as it is ambiguous.

devversion replied:
As discussed offline, we will be erroring if we find conflicts. Right now this shouldn't be the case, especially since we are better than AIO with that, given that "imports are detected via the local context of declarations". e.g. core/testing `inject` will be properly linked in test entry-points, while for non-testing, `core` `inject` may be used.

devversion reviewed (approved):
LGTM, great work! Minor comments in the compiler

JeanMeche commented on packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:
```diff
@@ -27,9 +27,13 @@ import {
 type DeclarationWithExportName = readonly [string, ts.Declaration];
 
+/** The modules here doesn't expose symbols publicly as they are considered private */
+const privateModules = new Set(['core/primitives/signals', 'core/primitives/event-dispatch']);
```
Understood, let's handle this in a follow-up !

AndrewKushnir commented:
@JeanMeche there is a conflict with the `main` branch. Could you please rebase and resolve the merge conflict when you get a chance? Thank you.

AndrewKushnir commented:
[Presubmit](http://test/OCL:668991633:BASE:668991673:1724953829423:833a5110).

AndrewKushnir commented:
This PR was merged into the repository by commit 01030d58a7162cda2fa597ac77e29ea7066b1a50.

The changes were merged into the following branches: main, 18.2.x
